db/src/master.rs

```diff
@@ -478,7 +478,7 @@ impl Master {
         // Lookup the tenant, and get a handle to the allocator. Required to avoid capturing a
         // reference to Master in the generator below.
         let tenant = self.get_tenant(tenant_id);
-        let alloc = self.heap.clone();
+        let alloc: *const Allocator = &self.heap;
```

Comment:
I still wonder about this. get() returns the gen, and because gen outlasts its enclosing scope it has to be move. But isn't there a reference type we can use? I've forgotten what the compiler complains about, but is the issue that everything that the closure captures has to be Sync and Send? Otherwise, why can't we just pass in a reference or something like Box?

For example, we could probably ensure (maybe we already do) that the data structure we hold the generators in is guaranteed to outline Master/Allocator. So, it seems like we can prove to the compiler that the lifetime is ok on alloc. But, is it that a ref on Master's alloc isn't Sync/Send that we run into problems?

Reply:
`Box<T>` is for single ownership.
`Rc<T>` is for multiple ownership.
`Arc<T>` is for multiple ownership, and threadsafe.

And compiler won't allow a reference other than `Arc` as all the threads share this reference. On the other hand, a raw pointer is neither `Send` and nor `Sync`, but we can share it among threads with the help of unsafe rust. We can't convert a raw pointer to a reference in the safe rust as there are no gurantees about its lifetime etc.